### ISSUE TITLE
Request: accept more parameters

### ISSUE THREAD

**Issue by r3incarnat0r** [Disassembler0/Win10-Initial-Setup-Script#81]
**TL;DR**: Please allow in parameters to include multiple presets, tweaks and functions from external file at the same time. :)

Hello and thanks for the awesome script!
I would like you to consider to extend capability of accepting multiple parameters and also ability to load external tweaks. That's more requests at the same time but I throw them here all to see the whole context:

1) Ability to add multiple presets.
2) Ability to add multiple separate tweaks together with presets.
3) Ability to include functions from separate file (e.g. with custom tweaks that are not suitable for upstream).
4) Process presets and separate tweak parameters in entered order.

So the example of possible run command could be something like:

`powershell.exe -NoProfile -ExecutionPolicy Bypass -File "Win10.ps1" -include "CustomWeirdTweaks.ps1" -preset "DefaultThatIAlwaysRun.preset" -preset "DiffForScenarioX.preset" -tweak "TuneSomethingSpecific EnableSomeTweakThisTime"`

This would (obviously) allow to:
- apply desired combination in one pass (and one command)
- keep specific diffs separated without editing base Win10.ps1
- easily update base Win10.ps1 just by replacing this file

Thanks!

**Comment by Disassembler0** [comment]
**tl;dr:** Already tried that, found it to be fragile and very difficult to implement.

I actually wanted to have similar functionality and even attempted to write a draft for it, including a feature which you haven't listed, which would strip a tweak from preset for current run. Something in the lines of

```
powershell.exe -File "Win10.ps1" -Preset "Some.preset" -Tweaks "Tweak1,-Tweak2"
```

where the minus sign in front of `Tweak2` would mean that the tweak wouldn't be applied even if it would be present in the preset.

---

Now for the bedtime story:

Turns out that the shell variety in Windows and their respective argument parsing properties make such functionality considerably difficult to implement. There are several problems which need to be solved in one go

- Differences in execution environment and cmd/PowerShell arguments processing
- Argument escaping, most notably spaces and commas in path names
- Re-running the script in elevated shell while retaining correctly escaped arguments
- Replacing relative file paths with absolute ones during shell elevation

Take this simple stupid script which is supposed to just echo back two command line parameters, one string and one array.

```
Param(
    [String]$stringparam,
    [String[]]$arrayparam
)

Write-Output "String param: $stringparam"
Write-Output "Array param:"
$arrayparam | ForEach { Write-Output $_ }
```

Depending on how you run it, the contents of the variables will differ. Observe the protection used (quotes vs apostrophes), separators (comma vs space) and number of elements in resulting output.

```
C:\Test>powershell -File test.ps1 -StringParam "One,Two Three" -ArrayParam "One,Two Three"
String param: One,Two Three
Array param:
One,Two Three
```
^ String is OK, Array is wrong.
```
C:\Test>powershell .\test.ps1 -StringParam "One,Two Three" -ArrayParam "One,Two Three"
String param: One Two
Array param:
One
Two
```
^ Everything is wrong due to stripping and implicit conversion to `System.Object[]`.
```
C:\Test>powershell -File test.ps1 -StringParam 'One,Two Three' -ArrayParam 'One,Two Three'
String param: 'One,Two
Array param:
'One,Two
```
^ ¯\\\_(ツ)\_/¯


```
C:\Test>powershell .\test.ps1 -StringParam 'One,Two Three' -ArrayParam 'One,Two Three'
String param: One,Two Three
Array param:
One,Two Three
```
^ See example no. 1.

The reason is that when you use `-File`, the arguments are passed as literals and are not parsed by PowerShell further. On the other hand, when you don't use `-File`, unnamed PowerShell argument defaults to `-Command`, so the whole string, including the arguments, is re-executed in PowerShell and the arguments are parsed for the second time there, thus possibly screwing up whatever was sane when it was passed from cmd. The only way how to pass the parameters without breaking anything is
```
C:\Test>powershell .\test.ps1 -StringParam 'One,Two Three' -ArrayParam 'One','Two Three'
String param: One,Two Three
Array param:
One
Two Three
```
This also leaves `[environment]::GetCommandLineArgs()` sane, which would normally be a good thing, but due to problems with possible relative file paths during elevation, it needs to be inspected and fixed anyway. Moreover, everybody would be forced to use *the only correct formatting* and you know how people are good in reading documentation and following rules (*Hint: They aren't*).

Additional problems arise when you need to take all the arguments which were passed to the script and forward them into elevated shell. This was problematic for long time even with the basic functionality which the script has now until I've worked around it in https://github.com/Disassembler0/Win10-Initial-Setup-Script/commit/d427ae735487e6fb35dec95b1d54c9f92c987ebd. In this particular case, I would need to retain the list of included scripts and the list of presets (and of course the list of tweaks, but that's a non-issue). Since the elevation is losing working directory, no relative paths will work in the elevated script anymore, so they need to be resolved before the elevation and they may contain additional special characters (spaces, commas) requiring further sanitizing.

So, the conclusion is that *if* I would implement such features, due to the problems described above, I would have to either force people to start using the script *the correct way* or I would be forced to write my own argument parser, both of which are generally bad ideas. Nevertheless I have learned a few new thing since the last time and I have some ideas about how to make it (probably) work, so when I will be really bored, I might give it another try. It will definitely need some extensive testing though.

**Comment by Hexalt** [comment]
Any plans to update this like BlackViper's service configuration?

https://github.com/madbomb122/BlackViperScript

**Comment by Disassembler0** [comment]
@Hexalt: Never heard of it. But if you mean to add crapload of command line switches, then no, not really. The presets should be the primary way of customization. I strive (or at least try to) for the simplicity and cleanliness of the code.

**Comment by r3incarnat0r** [comment]
If command line arguments are no-go, there's an alternative idea, but maybe it's still too crazy for the simple and clean code, though. ;) How about directives approach:
- arguments would be placed in preset files (basically functions like tweaks are)
- main script would contain "independent" functions, no need calling them by default, just if preset wishes so, just like tweaks
- processing $args much simpler - only one argument is enough for function (but could be called multiple times), also ordering should be solved - funcs would be called line by line as an user writes them.

Self-descriptive .preset example:

```
# make the script extra functions aware :)
IncludeModule Win10.CustomTweaks.psm1

# "dependencies":
# process these now
RequirePreset default.preset
RequirePreset finetune.preset

RequireAdmin

# additional Win10.ps1 standard tweaks:
DisableTelemetry
DisableWiFiSense

# custom tweaks from included module are available
EnableCustomTweak1
DisableCustomTweak2
```

**Comment by Disassembler0** [comment]
Command line arguments are OK, just not the amount and variety as the linked script has. (I actually didn't get @Hexalt's question, so I just assumed it's about the arguments, since this issue is about them.)

The complexity wouldn't probably differ that much if it would be an argument parser or a config file parser. The problems with argument sanitation and absolute path resolution during elevation would still need to be solved one way or another. And I already see few problems with the config-like approach which wouldn't exist with arguments. Eg. nesting of `RequirePreset` or even worse, circular references. Even though I like how clean the config looks like, I rather try my luck with args parser. :)

Btw. the order of tweaks really shouldn't matter. If it does, then the tweak should be rewritten or at least some checks should be added.

**Comment by r3incarnat0r** [comment]
OK, but are there any plans to implement the mentioned ability to include custom tweak functions from external files anyhow? No args or config needed - how about to auto include for example all .psm1 files contained in Modules subdir or let's say files such as Win10.*.Plugin.psm1 from main script's dir or something? Thanks for considering it... :)

**Comment by r3incarnat0r** [comment]
Recently I mentioned modules support would be nice, but it's not needed actually, since I found a quite simple native way how to accomplish it using just functionality already present. I realized a preset file is a list of expressions executed using Invoke-Expression, so the script can be easily and naturally extended with custom tweaks just by properly handling such preset file.

Example module file CustomTweaks.psm1:

```
function EnableHocusPocus  {
	Write-Output "Enabling Hocus Pocus..."
}

function DisableHocusPocus  {
	Write-Output "Disabling Hocus Pocus..."
}

Export-ModuleMember -Function 'Enable*'
Export-ModuleMember -Function 'Disable*'
```

Example preset file CustomTweaks.preset:

```
# Load Custom Tweaks
Import-Module .\CustomTweaks.psm1

# Custom Tweaks Ready To Roll Out!
EnableHocusPocus
DisableHocusPocus

# Aux
WaitForKey
```

PS: Now I just hope you won't consider this approach as a security flaw and block it, so I would miss this capability again. :D

**Comment by Disassembler0** [comment]
This is the most lazy approach to the whole thing, so this is the only way how I wanted do that. :)

It also allows to just "source" *.ps1 files with functions without a need for an actual PS module with export, so yes, this is how it's gonna work. But that's the easy part. Argument parsing will be the hard one. Not sure how soon I'll get to it, but I'm hoping for second half of May.

**Comment by Disassembler0** [comment]
@r3incarnat0r: You are hereby tasked with a quest of utmost importance. Test the [includes branch](https://github.com/Disassembler0/Win10-Initial-Setup-Script/tree/includes) and let me know if it fits your  needs. :)

Relevant part of the commit: https://github.com/Disassembler0/Win10-Initial-Setup-Script/commit/ae9814886e2510ff92fdc814c9a2df7a4dad7733#diff-ce2251917e3d8329e962acfbb02332a1

**Comment by Disassembler0** [comment]
Implemented in 1e977ab95fe04a99cdd152d342586de01d0aec96.
